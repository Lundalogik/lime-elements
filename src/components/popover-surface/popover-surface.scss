@use '../../style/functions';
@use '../../style/mixins';

:host(limel-popover-surface) {
    isolation: isolate;

    position: relative;
    display: flex;

    width: var(--popover-surface-width, auto);
    max-height: inherit; // inherits it from the dynamically calculated `max-height` of `limel-portal--container`
    max-width: calc(100vw - 2rem);
    margin: 0 0.25rem;
}

.limel-popover-surface {
    flex: 1;
    min-width: 0;
    min-height: 0;
    border-radius: var(--popover-border-radius, functions.pxToRem(12));
    box-shadow: var(--shadow-depth-16);

    backdrop-filter: blur(functions.pxToRem(5));
    -webkit-backdrop-filter: blur(functions.pxToRem(5));

    &:after {
        // allows using `--popover-body-background-color` while
        // getting the blurred backdrop effect
        transition: opacity 0.4s ease;
        pointer-events: none;

        content: '';
        position: absolute;
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
        z-index: -1;

        opacity: 0.75;

        border-radius: var(--popover-border-radius, functions.pxToRem(12));
        background-color: var(
            --popover-body-background-color,
            var(--lime-elevated-surface-background-color)
        );
    }

    @include mixins.visualize-keyboard-focus;

    &:focus,
    &:focus-within {
        &:after {
            opacity: 1;
        }
    }

    & > * {
        // this fixes some unwanted layout issues
        // which might be very hard for consumers to realize
        // and fix by themselves. Without this code,
        // if the consumer sets for example paddings in
        // what they send to the popover, their element
        // can grow larger than the popover surface
        box-sizing: border-box;
    }
}
