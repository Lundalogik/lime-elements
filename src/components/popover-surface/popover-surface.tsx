import { Component, h, Prop, Element } from '@stencil/core';

/**
 * @slot - Content to put inside the surface
 * @private
 */

@Component({
    tag: 'limel-popover-surface',
    shadow: true,
    styleUrl: 'popover-surface.scss',
})
export class PopoverSurface {
    /**
     * Content to render
     */
    @Prop()
    public contentCollection: HTMLCollection;

    @Element()
    private host: HTMLLimelPopoverSurfaceElement;

    public componentDidLoad() {
        this.appendElement();
    }

    public render() {
        return <div class="limel-popover-surface" tabindex="0" />;
    }

    private appendElement() {
        const portalContainer = this.host.shadowRoot.querySelector(
            '.limel-popover-surface',
        );

        Array.from(this.contentCollection).forEach((child) => {
            if (child.slot === 'trigger') {
                return;
            }

            portalContainer.appendChild(child);
        });
    }
}
