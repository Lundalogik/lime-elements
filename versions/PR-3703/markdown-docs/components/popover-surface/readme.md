# limel-popover-surface



<!-- Auto Generated Below -->


## Properties

| Property            | Attribute | Description       | Type             | Default     |
| ------------------- | --------- | ----------------- | ---------------- | ----------- |
| `contentCollection` | --        | Content to render | `HTMLCollection` | `undefined` |


## Slots

| Slot | Description                       |
| ---- | --------------------------------- |
|      | Content to put inside the surface |


## Dependencies

### Used by

 - [limel-popover](../popover)

### Graph
```mermaid
graph TD;
  limel-popover --> limel-popover-surface
  style limel-popover-surface fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
