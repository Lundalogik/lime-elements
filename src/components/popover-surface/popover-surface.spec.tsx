import { newSpecPage, SpecPage } from '@stencil/core/testing';
import { h } from '@stencil/core';
import { PopoverSurface } from './popover-surface';

let page: SpecPage;

test('render component without children', async () => {
    await createComponent([]);

    expect(page.root).toMatchSnapshot();
});

test('render component with one child', async () => {
    const child1 = document.createElement('div');
    child1.textContent = 'Child 1';

    await createComponent([child1]);

    expect(page.root).toMatchSnapshot();
});

test('render component with two children', async () => {
    const child1 = document.createElement('div');
    const child2 = document.createElement('div');
    child1.textContent = 'Child 1';
    child2.textContent = 'Child 2';

    await createComponent([child1, child2]);

    expect(page.root).toMatchSnapshot();
});

test('render component with trigger element', async () => {
    const child1 = document.createElement('div');
    child1.textContent = 'Child 1';
    child1.slot = 'trigger';

    await createComponent([child1]);

    expect(page.root).toMatchSnapshot();
});

async function createComponent(children: HTMLElement[]): Promise<void> {
    const aside = document.createElement('aside');
    document.body.append(aside);
    for (const child of children) {
        aside.append(child);
    }

    page = await newSpecPage({
        components: [PopoverSurface],
        template: () => (
            <limel-popover-surface contentCollection={aside.children} />
        ),
    });

    await page.waitForChanges();
}
