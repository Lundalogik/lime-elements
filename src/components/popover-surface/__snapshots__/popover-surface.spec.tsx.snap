// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render component with one child 1`] = `
<limel-popover-surface>
  <mock:shadow-root>
    <div class="limel-popover-surface" tabindex="0">
      <div>
        Child 1
      </div>
    </div>
  </mock:shadow-root>
</limel-popover-surface>
`;

exports[`render component with trigger element 1`] = `
<limel-popover-surface>
  <mock:shadow-root>
    <div class="limel-popover-surface" tabindex="0"></div>
  </mock:shadow-root>
</limel-popover-surface>
`;

exports[`render component with two children 1`] = `
<limel-popover-surface>
  <mock:shadow-root>
    <div class="limel-popover-surface" tabindex="0">
      <div>
        Child 1
      </div>
      <div>
        Child 2
      </div>
    </div>
  </mock:shadow-root>
</limel-popover-surface>
`;

exports[`render component without children 1`] = `
<limel-popover-surface>
  <mock:shadow-root>
    <div class="limel-popover-surface" tabindex="0"></div>
  </mock:shadow-root>
</limel-popover-surface>
`;
